api/v1alpha3/machinehealthcheck_webhook.go

```diff
@@ -0,0 +1,100 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+func (m *MachineHealthCheck) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(m).
+		Complete()
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-cluster-x-k8s-io-v1alpha3-machinehealthcheck,mutating=false,failurePolicy=fail,groups=cluster.x-k8s.io,resources=machinehealthchecks,versions=v1alpha3,name=validation.machinehealthcheck.cluster.x-k8s.io
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-cluster-x-k8s-io-v1alpha3-machinehealthcheck,mutating=true,failurePolicy=fail,groups=cluster.x-k8s.io,resources=machinehealthchecks,versions=v1alpha3,name=default.machinehealthcheck.cluster.x-k8s.io
+
+var _ webhook.Defaulter = &MachineHealthCheck{}
+var _ webhook.Validator = &MachineHealthCheck{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (m *MachineHealthCheck) Default() {
+	PopulateDefaultsMachineHealthCheck(m)
+}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (m *MachineHealthCheck) ValidateCreate() error {
+	return m.validate(nil)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (m *MachineHealthCheck) ValidateUpdate(old runtime.Object) error {
+	mhc, ok := old.(*MachineHealthCheck)
+	if !ok {
+		return fmt.Errorf("expected a MachineHealthCheck but got a %T", old)
+	}
+	return m.validate(mhc)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (m *MachineHealthCheck) ValidateDelete() error {
+	return nil
+}
+
+func (m *MachineHealthCheck) validate(old *MachineHealthCheck) error {
+	var allErrs field.ErrorList
+
+	// Validate selector parses as Selector
+	_, err := metav1.LabelSelectorAsSelector(&m.Spec.Selector)
+	if err != nil {
+		allErrs = append(
+			allErrs,
+			field.Invalid(field.NewPath("spec", "selector"), m.Spec.Selector, err.Error()),
+		)
+	}
+
+	if old != nil && old.Spec.ClusterName != m.Spec.ClusterName {
+		allErrs = append(
+			allErrs,
+			field.Invalid(field.NewPath("spec", "clusterName"), m.Spec.ClusterName, "field is immutable"),
+		)
+	}
+
+	if len(allErrs) == 0 {
+		return nil
+	}
+
+	return apierrors.NewInvalid(GroupVersion.WithKind("MachineHealthCheck").GroupKind(), m.Name, allErrs)
+}
+
+// PopulateDefaultsMachineHealthCheck fills in default field values.
+// This is also called during MachineHealthCheck sync.
```

Comment:
This comment says that this method is called during "MachineHealthCheck sync", is that accurate? I don't see any additional uses of it.

Reply:
I copied it over from the Defaulter that I based this on. I assumed it meant that it would be called during updates to the object caused by the MachineHealthCheck sync... I think it's confusing, I'll remove it 😅